### PR TITLE
Refactor: Allow More Answer Types to Be Cacheable

### PR DESCRIPTION
**Pull request by developStorm**
This PR introduces the `WithBaseAnswer` interface as the cache value type to allow broader answer types to be included in the cache. The `answers_generate.go` script, when run with `go generate`, implements the required interface function `BaseAns` for each eligible answer struct.

### CODE REVIEW

File: src/zdns/answers_generate.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
+//go:build ignore
```
Comment (phillip-stephens) [inline src/zdns/answers_generate.go:1]: Let's add the license and attribution linking to the miekg project.

File: src/zdns/cache.go [hunk 2]
```diff
@@ -169,68 +169,68 @@ func (s *Cache) getCachedResult(q Question, ns *NameServer, isAuthority bool, de
...
 }
 
 func (s *Cache) buildCachedResult(res *SingleQueryResult, depth int, layer string) *CachedResult {
 	now := time.Now()
 	cachedRes := CachedResult{}
 	cachedRes.Answers = make([]TimedAnswer, 0, len(res.Answers))
-	for _, a := range res.Answers {
-		castAns, ok := a.(Answer)
+
+	var getExpirationForSafeAnswer = func(a any) (WithBaseAnswer, time.Time) {
```
Comment (phillip-stephens) [inline src/zdns/cache.go:191]: I like this helper fn, cleans up things towards the end of the function!

**Review by phillip-stephens — COMMENTED** [review]
LGTM once we add the license stuff

**Comment by phillip-stephens** [comment]
And for context, this is going to enable us to cache `DNSKEY` and `DS` entries so @developStorm can add DNSSEC validation support into ZDNS.

File: src/zdns/answers_generate.go (hunk 1 above)
Comment (developStorm) [reply src/zdns/answers_generate.go:1]: added in https://github.com/zmap/zdns/pull/465/commits/6f09067a5b2b0e87abf4163ec2c12709f85bb37f

**Review by dadrian — APPROVED** [review]

**Pull request merged**